t/python/t0015-job-output.py

```diff
@@ -0,0 +1,600 @@
+        def watch_cb(future, result):
+            try:
+                stream, data = future.get_output()
+            except Exception as exc:
+                stream = "stderr"
+                data = str(exc)
+                future.reset()
+            result.setdefault(stream, []).append(data)
+
+        def get_results(result):
+            return {
+                key: "".join(result.get(key, [])) for key in ("stdout", "stderr", "log")
+            }
+
+        results = {}
+
+        # Basic
+        results["basic"] = {}
+        output_watch_async(self.fh, self.submit()).then(watch_cb, results["basic"])
+
+        # Separate logs
+        results["log"] = {}
+        output_watch_async(
+            self.fh, self.submit(verbose=2), log_stderr_level=LOG_QUIET
+        ).then(watch_cb, results["log"])
+
+        # Labelio
+        results["labelio"] = {}
+        output_watch_async(self.fh, self.submit(ntasks=2), labelio=True).then(
+            watch_cb, results["labelio"]
+        )
+
+        # Cancel
+        jobid = self.submit(hold=True)
+        flux.job.cancel(self.fh, jobid)
+        results["cancel"] = {}
+        output_watch_async(self.fh, jobid).then(watch_cb, results["cancel"])
+
+        self.fh.reactor_run()
+
+        for test in ("basic", "log", "labelio", "cancel"):
+            results[test] = get_results(results[test])
+
+        self.assertEqual(results["basic"]["stdout"], self.test_stdout)
+        self.assertEqual(results["basic"]["stderr"], self.test_stderr)
+        self.assertEqual(results["basic"]["log"], "")
+
+        self.assertEqual(results["log"]["stdout"], self.test_stdout)
+        self.assertEqual(results["log"]["stderr"], self.test_stderr)
+        self.assertTrue(len(results["log"]["log"]) > 0)
+
+        self.assertIn("0: line 1", results["labelio"]["stdout"])
+        self.assertIn("1: line 1", results["labelio"]["stdout"])
+        self.assertIn("0: error 1", results["labelio"]["stderr"])
+        self.assertIn("1: error 1", results["labelio"]["stderr"])
+
+        # Also, should have received a job.exception message on stderr:
+        self.assertIn("job.exception", results["cancel"]["stderr"])
+        self.assertIn("never started", results["cancel"]["stderr"])
+
+    def test_output_watch_async_exception(self):
+        def watch_cb_with_error(future):
+            # Test raising error from typo:
+            stream, line = notfuture.getline()  # noqa: F821
+
+        jobid = self.submit()
+        output_watch_async(self.fh, jobid).then(watch_cb_with_error)
+        with self.assertRaises(NameError):
+            self.fh.reactor_run()
+
+    def test_output_watch_lines(self):
+        def do_watch_lines(
+            jobid,
+            result=None,
+            label=False,
+            log_stderr_level=LOG_TRACE,
+            keepends=False,
+            nojoin=False,
+            nowait=False,
+        ):
+            if result is None:
+                result = {}
+            count = 0
```

Comment:
just thought about this post my PR approval, separate stdout_count and stderr_count?  Probably not necessary since you check stdout/stderr output below to ensure they match.   Just to be thorough and would allow the additional count check below that you skip b/c of the potential for log output.

Reply:
Not a bad idea. Maybe I'll try it to see how messy it is. There is one benefit to checking an overall count though -- it would catch cases where we're getting callbacks we don't expect where stream isn't `stderr` or `stdout`... maybe I'll just leave as is for now.

There is another gotcha with the `*lines` watchers -- if a partial line exists within a given eventlog `data` event, that partial will be treated as a whole line currently. That is, the `JobOutputWatchLines` class doesn't currently buffer lines, so really this is kind of "best effort".

This could be fixed in the future if necessary, or we could drop the "lines" based watchers if that shortcoming is a showstopper. (callers could always use `str.splitlines()` if they want to do something per line), though I figured it is maybe acceptable for now (I think most output is attempted to be line buffered at the job shell).

Because of that, I'm not sure pedantic line counting is worth the effort at this point.

Reply:
> There is another gotcha with the *lines watchers -- if a partial line exists within a given eventlog data event, that partial will be treated as a whole line currently. That is, the JobOutputWatchLines class doesn't currently buffer lines, so really this is kind of "best effort".

Is this documented in the class?  Didn't see an obvious mention of it (on phone so could has missed) but might be good to add.

Reply:
Yeah, I was still wondering if the watch by lines class should instead just be dropped for now? Unsure if there's a compelling use case at this point? If you think it is fine to keep with the comments, then I'm fine leaving it in too.

Reply:
I'd say leave it in, I think it's a "net win" for now.  If it becomes a problem, could enhance the class later.

Reply:
Ok, that sounds good to me. I'll add some caveats to the docstrings